rwkv.h

```diff
@@ -83,7 +83,8 @@ extern "C" {
     // Returns NULL on any error. Error messages would be printed to stderr.
     // - model_file_path: path to model file in ggml format.
     // - n_threads: count of threads to use, must be positive.
-    RWKV_API struct rwkv_context * rwkv_init_from_file(const char * model_file_path, const uint32_t n_threads);
+    // - n_gpu_layer: count of layers need to load to gpu (only works when cuBLAS is on)
+    RWKV_API struct rwkv_context * rwkv_init_from_file(const char * model_file_path, const uint32_t n_threads, const uint32_t n_gpu_layers);
```

Comment:
I'm not sure I'm comfortable with changing function signature. In the past, changing signatures lead to users with out of date libraries complaining about `rwkv.cpp` not working.

At the same time, I don't have a really good idea how we can extend C API without breaking compatibility. Maybe adding new functions like `rwkv_init_from_file_ex`, `rwkv_init_from_file_ex_ex` is the way lol

@LoganDark Since you are knowledgable in C/C++, would you like to comment on this? :)

Reply:
> @LoganDark Since you are knowledgable in C/C++, would you like to comment on this? :)

there are two strategies you could go for:

- A**P**I compatibility: define the old function signature in the header file as a shortcut that fills in a default of gpu_layers (possibly 0).

- A**B**I compatibility: define the old function signasture in the **cpp file** and not the header file. the built DLL will contain it, and old apps that do not know about the new argument will work

you most likely want ABI compatibility (new dlls work with old executables), not API compatibility (new source code works with old source code). so define the old function in cpp that returns a call to the new signature.

Reply:
Thanks for clarification!

I think I'm concerned with a specific scenario: built library (dll/so) is *old*, byt Python code is *new*. Example of how this may happen: a user clones `rwkv.cpp` and downloads prebuilt dll, `rwkv.cpp` updates, user pulls; but forgets to re-download the dll.

In the past, I've replaced `int32` argument with `char *` argument, which lead to old dll receiving trash values into the `int` argument.

In this case, if I understand correctly, int value for `gpu_layers` would be put into some register by Python, and old dll will just ignore it. It needs to be actually tested whether this is the case. If it is, then I'm okay with adding the parameter; but, as @LoganDark proposed in "ABI compatibility section", keep function `rwkv_init_from_file(model_file_path, n_threads)` in cpp file that just calls the actual function passing `0` for `gpu_layers`.

Reply:
the real solution (to scale to any amount of parameters) is to switch to key value lists. this is possible with  (afaik) `const char*[2][]` in the form of `{{key1, value1}, {key2, value2}, {NULL, NULL}}`. once I am done with my PR draft I may be able to help out more with this.

Reply:
added rwkv_init_from_file(model_file_path, n_threads) in the cpp file for now, may update by using other parameter list in the future